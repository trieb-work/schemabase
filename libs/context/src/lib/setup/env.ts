import { ExtendContextFn } from "../context"

import { env } from "@eci/util/env"

export type ServiceConfig = {
  redis: {
    host: string
    password: string
    port: number
  }
  oauth: {
    google: {
      clientId: string
      clientSecret: string
    }
  }
  elasticSearch: {
    apmServer: string
    apmSecretToken: string
    loggingServer: string
  }
}

/**
 * Fetch the client's configuration and expose it to the context
 */
export const setupConfig = (): ExtendContextFn<"serviceConfig"> => async (ctx) => {
  const serviceConfig: ServiceConfig = {
    redis: {
      host: env.require("REDIS_HOST"),
      password: env.require("REDIS_PASSWORD"),
      port: parseInt(env.require("REDIS_PORT"), 10),
    },
    oauth: {
      google: {
        clientId: env.require("GOOGLE_OAUTH_CLIENT_ID"),
        clientSecret: env.require("GOOGLE_OAUTH_CLIENT_SECRET"),
      },
    },
    elasticSearch: {
      apmServer: env.require("ELASTIC_APM_SERVER"),
      apmSecretToken: env.require("ELASTIC_APM_SECRET_TOKEN"),
      loggingServer: env.require("ELASTIC_LOGGING_SERVER"),
    },
  }
  ctx.serviceConfig = serviceConfig
  return ctx
}
