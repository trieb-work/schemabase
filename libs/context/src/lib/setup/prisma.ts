import { ExtendContextFn } from "../context"
import { PrismaClient } from "@eci/data-access/prisma"

/**
 * Initialize a prisma client and make it public to the context
 */
export const setupPrisma = (): ExtendContextFn<"prisma"> => async (ctx) => {
  ctx.prisma = new PrismaClient()
  return ctx
}
