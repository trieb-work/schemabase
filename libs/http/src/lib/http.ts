// eslint-disable-next-line no-restricted-imports
import axios from "axios";
import { ECI_TRACE_HEADER } from "@eci/util/constants";

export type Request = {
  method: "GET" | "POST" | "PUT" | "DELETE";
  url: string;
  body?: unknown;
  headers?: Record<string, string | number>;
  params?: Record<string, string | number>;
};

export type Response<Data> = {
  headers: Record<string, string | number>;
  status: number;
  data: Data | null;
};

export type HttpApi = {
  call: <Data>(req: Request) => Promise<Response<Data>>;
  setHeader(name: string, value: string | number): void;
};

export class HttpClient implements HttpApi {
  public headers: Record<string, string | number>;

  public constructor(config?: { traceId?: string }) {
    this.headers = {};
    if (config?.traceId) {
      this.setHeader(ECI_TRACE_HEADER, config?.traceId);
    }
  }

  public setHeader(name: string, value: string | number): void {
    this.headers[name] = value;
  }

  public async call<Data>(req: Request): Promise<Response<Data>> {
    return await axios({
      method: req.method,
      url: req.url,
      params: req.params,
      headers: { ...this.headers, ...req.headers },
      data: JSON.stringify(req.body),
    }).then((res) => ({
      status: res.status,
      data: res.data ?? null,
      headers: res.headers,
    }));
  }
}
