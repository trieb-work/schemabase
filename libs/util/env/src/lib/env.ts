import { EnvironmentVariableNotFoundError } from "@eci/util/errors"
/**
 * Load an environment variable.
 * Returns the fallback if not found.
 */
export function loadEnv(key: string, fallback?: string): string | undefined {
  const value = process.env[key]
  return value ?? fallback
}

/**
 * Load an environment variable or throw an error if not found.
 */
export function requireEnv(key: string): string {
  const value = process.env[key]
  if (!value) {
    throw new EnvironmentVariableNotFoundError(key)
  }
  return value
}
