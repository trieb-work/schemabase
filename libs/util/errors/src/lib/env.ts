import { GenericError } from "./base";

export class EnvironmentVariableNotFoundError extends GenericError {
  /**
   * @param name - The name of the environment variable
   */
  constructor(name: string) {
    super(
      "EnvironmentNotFoundError",
      `Environment variable "${name}" not found.`,
      {},
    );
  }
}
