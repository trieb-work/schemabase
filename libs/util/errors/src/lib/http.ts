import { GenericError } from "./base";

/**
 * HttpError is used to pass a status code inside nextjs api routes.
 */
export class HttpError extends GenericError {
  public readonly statusCode: number;
  constructor(statusCode: number, message = "HttpError") {
    super("HttpError", message);
    this.statusCode = statusCode;
  }
}
