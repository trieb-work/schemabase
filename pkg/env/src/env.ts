export class EnvironmentVariableNotFoundError extends Error {
  public readonly variableName: string;
  /**
   * @param name - The name of the environment variable.
   */
  constructor(name: string) {
    super(`Environment variable "${name}" not found.`);
    this.variableName = name;
  }
}
/**
 * Load an environment variable.
 * Returns the fallback if not found.
 */
export function getEnv(key: string, fallback?: string): string | undefined {
  return process.env[key] ?? fallback;
}

/**
 * Load an environment variable or throw an error if not found.
 * @throws EnvironmentVariableNotFoundError
 */
export function requireEnv(key: string): string {
  const value = process.env[key];
  if (!value) {
    throw new EnvironmentVariableNotFoundError(key);
  }
  return value;
}
