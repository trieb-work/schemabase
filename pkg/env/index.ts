import { getEnv, requireEnv } from "./src/env";
export { EnvironmentVariableNotFoundError } from "./src/env";

export const env = {
  get: getEnv,
  require: requireEnv,
};
