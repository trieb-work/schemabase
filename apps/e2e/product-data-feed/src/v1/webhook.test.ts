import { HttpClient } from "@eci/http";
import { PrismaClient } from "@eci/data-access/prisma";
import { createHash } from "crypto";
import {idGenerator} from "@eci/util/ids"
import {
  CountryCode,
  createSaleorClient,
  ProductTypeKindEnum,
} from "@eci/adapters/saleor/api";
import { env } from "@chronark/env";

const webhookId = idGenerator.id("test");

beforeAll(async () => {
  const prisma = new PrismaClient();

  let saleorClient = await createSaleorClient({
    traceId: idGenerator.id("trace"),
    graphqlEndpoint: env.require("SALEOR_URL"),
  });

  const setup = await saleorClient.tokenCreate({
    email: "admin@example.com",
    password: "admin",
  });
  if (!setup?.tokenCreate?.token) {
    throw new Error(`Unable to get saleor token`);
  }

  saleorClient = createSaleorClient({
    traceId: idGenerator.id("trace"),
    graphqlEndpoint: env.require("SALEOR_URL"),
    token: setup.tokenCreate.token,
  });

  const categoryResponse = await saleorClient.categoryCreate({
    input: {
      name: "juices",
    },
  });
  const category = categoryResponse?.categoryCreate?.category;

  if (!category) {
    throw new Error(
      `Unable to create category: ${JSON.stringify(categoryResponse, null, 2)}`,
    );
  }

  const channelResponse = await saleorClient.channelCreate({
    input: {
      isActive: true,
      name: idGenerator.id("test"),
      slug: idGenerator.id("test"),
      currencyCode: "EUR",
      defaultCountry: CountryCode.De,
    },
  });
  const channel = channelResponse?.channelCreate?.channel;

  if (!channel) {
    throw new Error(
      `Unable to create channel: ${JSON.stringify(channelResponse, null, 2)}`,
    );
  }

  const productTypeResponse = await saleorClient.productTypeCreate({
    input: {
      name: idGenerator.id("test"),
      slug: idGenerator.id("test"),
      hasVariants: false,
      isDigital: true,
      isShippingRequired: false,
      weight: 2,
      kind: ProductTypeKindEnum.Normal,
    },
  });
  const productType = productTypeResponse?.productTypeCreate?.productType?.id;
  if (!productType) {
    throw new Error(
      `Unable to create productType: ${JSON.stringify(
        productTypeResponse,
        null,
        2,
      )}`,
    );
  }

  const productResponse = await saleorClient.productCreate({
    input: {
      name: "Apple Juice",
      slug: "apple-juice",
      category: category.id,
      // description: "Apple juice is great",
      // weight: 2,
      productType,
    },
  });
  const product = productResponse?.productCreate?.product;
  if (!product) {
    throw new Error("no product " + JSON.stringify(productResponse, null, 2));
  }
  const productVariantResponse = await saleorClient.productVariantCreate({
    input: {
      attributes: [],
      product: product.id,
      sku: "apple-juice-sku",
      trackInventory: true,
    },
  });
  const productVariant =
    productVariantResponse?.productVariantCreate?.productVariant;
  if (!productVariant) {
    throw new Error(
      "no productVariant " + JSON.stringify(productVariantResponse, null, 2),
    );
  }

  await saleorClient.productChannelListingUpdate({
    id: product.id,
    input: {
      updateChannels: [
        {
          channelId: channel.id,
          availableForPurchaseDate: null,
          isPublished: true,
          visibleInListings: true,
          isAvailableForPurchase: true,
          addVariants: [productVariant.id],
        },
      ],
    },
  });

  await saleorClient.productVariantChannelListingUpdate({
    id: productVariant.id,
    input: [
      {
        channelId: channel.id,
        price: "1",
        costPrice: "2",
      },
    ],
  });

  const tenant = await prisma.tenant.create({
    data: {
      id: idGenerator.id("test"),
      name: idGenerator.id("test"),
    },
  });

  /**
   * Upserting because there is a unique constraint on the domain
   */
  const saleorApp = await prisma.saleorApp.create({
    data: {
      id: idGenerator.id("test"),
      name: idGenerator.id("test"),
      tenantId: tenant.id,
      channelSlug: channel.slug,
      domain: "http://saleor.eci:8000",
    },
  });
  const productDataFeed = await prisma.productDataFeedApp.create({
    data: {
      id: idGenerator.id("test"),
      tenant: {
        connect: {
          id: tenant.id,
        },
      },
      productDetailStorefrontURL: "www.storefront.com",
      saleorApp: {
        connect: { id: saleorApp.id },
      },
    },
  });

  const secret = idGenerator.id("test");
  const secretHash = createHash("sha256").update(secret).digest("hex");
  await prisma.incomingProductDataFeedWebhook.create({
    data: {
      id: webhookId,
      productDataFeedApp: {
        connect: {
          id: productDataFeed.id,
        },
      },
      secret: {
        create: {
          id: secret,
          secret: secretHash,
        },
      },
    },
  });

  const integration = await prisma.productDataFeedIntegration.create({
    data: {
      id: idGenerator.id("test"),
      tenantId: tenant.id,
      enabled: true,
      productDataFeedAppId: productDataFeed.id,
      saleorAppId: saleorApp.id,
    },
  });
  await prisma.subscription.create({
    data: {
      id: idGenerator.id("test"),
      tenantId: tenant.id,
      payedUntil: new Date(Date.now() + 1000 * 60 * 5),
      productDataFeedIntegration: {
        connect: {
          id: integration.id,
        },
      },
    },
  });
  await prisma.$disconnect();
});

describe("productdatafeed", () => {
  const variants = ["facebookcommerce"]; //, "googlemerchant"];

  const http = new HttpClient();
  for (const variant of variants) {
    it(`generates a feed for ${variant}`, async () => {
      const res = await http.call<string>({
        method: "GET",
        url: `http://localhost:3000/api/product-data-feed/v1/${variant}/${webhookId}`,
      });
      expect(res.status).toBe(200);
      expect(res.headers["content-type"]).toEqual("text/csv");
      expect(res.headers["cache-control"]).toEqual(
        "s-maxage=1, stale-while-revalidate",
      );
      expect(res.data).toMatchSnapshot();
    }, 20_000);
  }
});
