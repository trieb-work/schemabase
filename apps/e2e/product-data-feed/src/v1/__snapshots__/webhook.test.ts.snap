// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`productdatafeed generates a feed for facebookcommerce 1`] = `
"id,title,description,rich_text_description,image_link,additional_image_link,link,price,sale_price,condition,gtin,brand,unit_pricing_measure,availability,google_product_category
apple-juice-sku,Apple Juice,,,,,www.storefront.comapple-juice,1 EUR,1 EUR,new,,undefined,,out of stock,
"
`;
